Match/exclude destination rules (#276)
tfpauly/privacy-proxy: draft-ietf-intarea-proxy-config.md

| proxy | No | String containing the URI template or hostname and port of the proxy, depending on the format defined by the protocol | String | "https://proxy.example.org:4443/masque{?target_host,target_port}" |
| alpn | Yes | An array of Application-Layer Protocol Negotiation protocol identifiers | Array of Strings | ["h3","h2"] |
| mandatory | Yes | An array of optional keys that client must understand and process to use this proxy | Array of Strings | ["match"] |

> | identifier | No | A string used to refer to the proxy, which can be referenced by other dictionaries | String | "udp-proxy" |

Reviewer: is the "identifier" key optional or not? Here it is marked as not optional, but the description says it is optional.

Author: ```suggestion
| identifier | No | A string used to refer to the proxy, which can be referenced by other dictionaries, such as `proxy-match`  | String | "udp-proxy" |
```

only keys that are registered in an IANA registry, defined in {{proxy-info-iana}} and that are marked
as optional.  The `mandatory` list MUST NOT include any entries that are not present in the sub-dictionary.

The value of `identifier` key is an optional string that can be used to refer to the
proxy from other dictionaries, specifically those defined in {{destinations}}. The
string value is an arbitrary JSON string. Identifier values MAY be duplicated
across different proxy dictionaries in the `proxies` array, which would indicate
that all references from other dictionaries to a particular identifier value applies

> to all matching proxies.

Reviewer 2: I think it's worth spelling out here that a proxy without `identifier` is expected to accept any request since it cannot appear in `proxy-match` or `proxy-exclude` dictionaries.

Reviewer: I agree with this.

> to all matching proxies.

Reviewer 2: ```suggestion
to all matching proxies. Proxies without `identifier` key are expected to accept any traffic since their destinations
cannot be contained in `proxy-match` array defined in {{destinations}}.
```

  "proxies": [
    {
      "protocol": "http-connect",

>       "proxy": "proxy.example.org:80",

Reviewer 2: Extra comma. Were you planning to add `identifier`?

Author: ```suggestion
      "proxy": "proxy.example.org:80"
```

a proxy. Destination restrictions are often used in VPN tunnel configurations such as split
DNS in IKEv2 {{?IKEV2SPLIT=RFC8598}}.

PvD Additional Information can be used to indicate that a set of proxies only allows access to
a limited set of destinations.

To support determining which traffic is supported by different proxies, this document defines
two new PvD Additional Information keys: `proxy-match` and `proxy-exclude`. Both keys have
values that are arrays of dictionaries, where each subdictionary describes a rule for
matching or excluding traffic from one or more proxies. These subdictionaries are referred
to as "destination rules", since they define rules about which destinations can be accessed
for a particular proxy.

| JSON Key | Optional | Description | Type |
| --- | --- | --- | --- |
| proxy-match | Yes | An array of destination rules defining targets accessible over this proxy | Array of Objects |
| proxy-exclude | Yes | An array of destination rules defining targets that cannot be accessed over this proxy | Array of Objects |

## Destination rule keys

> This document defined three keys for destination rules. Any destination rule in the `proxy-match` array
> MUST contain the `proxies` key with at least one proxy `identifier`. Any destination rule in the
> `proxy-exclude` array MAY contain the `proxies` key, which if present MUST contain at least one proxy `identifier`.

Reviewer 2: ```suggestion
This document defines four keys for destination rules. Any destination rule  MUST contain the `proxies` key. 
Value corresponding to the `proxies` key may be either an empty array which implies that no proxy defined in this
PvD can process matching traffic or an array with at least one proxy `identifier`. 
```

> MUST contain the contain the `proxies` key with at least one proxy `identifier`. Any destination rule in the

Reviewer 2: ```suggestion
MUST contain the `proxies` key with at least one proxy `identifier`. Any destination rule in the
```

All destination rules MUST also contain at least one other key use to describe the destination properties.
Each key MUST correspond to an array with at least one entry.

Extensions or proprietary deployments can define new keys to describe destination properties. Any destination
rules that include keys not known to the client, or values that cannot be parsed, MUST be ignored in their entirety.

| JSON Key | Optional | Description | Type | Example |

> | --- | --- | --- | --- |
> | proxies | An array of strings that match `identifier` values from the top-level `proxies` array | Array of Strings | ["tcp-proxy", "udp-proxy"] |
> | domains | An array of FQDNs and wildcard DNS domains | Array of Strings | ["www.example.com", "*.internal.example.com"] |
> | subnets | An array of IPv4 and IPv6 addresses and subnets | Array of Strings | ["2001:DB8::1", "192.168.1.0/24"] |
> | ports | An array of TCP and UDP port ranges | Array of Strings | ["80", "443", "1024-65535"] |

Reviewer 2: ```suggestion
| --- | --- | --- | --- | --- |
| proxies | No | An array of strings that match `identifier` values from the top-level `proxies` array | Array of Strings | ["tcp-proxy", "udp-proxy"] |
| domains | Yes | An array of FQDNs and wildcard DNS domains | Array of Strings | ["www.example.com", "*.internal.example.com"] |
| subnets | Yes | An array of IPv4 and IPv6 addresses and subnets | Array of Strings | ["2001:DB8::1", "192.168.1.0/24"] |
| ports | Yes | An array of TCP and UDP port ranges | Array of Strings | ["80", "443", "1024-65535"] |
```

> The `domains` array includes specific FQDNs and zones that are either accessible (for `proxy-match` 

Reviewer 2: ```suggestion
The `domains` array includes specific FQDNs and zones that are either accessible (for `proxy-match`
```

Trailing whitespace

> destination rules) or non-accessible (for `proxy-exclude` rules) using the proxy.

Author: Let's have this be an empty array

A wildcard prefix (`*.`) is used to indicate matching entire domains or subdomains instead of
specific hostnames. Note that this can be used to match multiple levels of subdomains. For example "*.example.com"
matches "internal.example.com" as well as "www.public.example.com".
Entries that include the wildcard prefix also SHOULD be treated as if they match
an FQDN that only contains the string after the prefix, with no subdomain. So,
an entry "*.example.com" in the `domains` array of a `proxy-match` rule would match the FQDN "example.com",
unless "example.com" was specifically included in the `domains` array of an `proxy-exclude` rule. This is
done to prevent commonly needing to include both "*.example.com" and "example.com"
in the `domains` array of a `proxy-match` rule.

The `subnets` array includes IPv4 and IPv6 address literals, as well as IPv4 and IPv6 address subnets

> written using CIDR notation. Entries in `subnets` array of a `proxy-match` rule correspond to IP addresses
> and subnets that are available through the proxy, while entries in `subnets` array of an `proxy-exclude`
> rule define IP addresses and subnets that ought not to be used with the proxy. Subnet-based destination

Reviewer 2: ```suggestion
written using CIDR notation. Subnet-based destination
```

information are only meant be used when applications are communicating with destinations identified by
only an IP address, and not a hostname.

Destination rules SHOULD NOT contain both the `domains` key and the `subnets` key. Many
clients will not resolve a domain being accessed through the proxy to an IP address before using
a proxy, so the subnet information may not be available.

The `ports` array includes specific ports (used for matching TCP and/or UDP ports), as well as
ranges of ports written with a low port value and a high port value, with a `-` in between.
For example, "1024-2048" matches all ports from 1024 to 2048, including the 1024 and 1028.

> If `ports` key is not present in a `proxy-match` or `proxy-exclude` object, all ports are assumed to match.

Reviewer 2: ```suggestion
If `ports` key is not present, all ports are assumed to match.
```

Comma-separated port list may contain individual port numbers (such as "80")
or inclusive ranges of ports. For example "1024-2048" matches all ports from 1024 to 2048, including the 1024 and 1028.

## Using destination rules

The destination rules described in `proxy-match` and `proxy-exclude` arrays can be used
to determine which traffic can be sent through proxies, and which specific set of proxies to
use for any particular connection. By evaluating the rules in order, a consistent behavior
for usage can be achieved.

> Rules in the `proxy-match` list SHOULD be provided in order of priority, such that a client

Reviewer 2: Is there a reason for a SHOULD here? SHOULD normally requires an explanation for the exception logic. I suggest going with a "MUST"

Author: Conversely, if this is a MUST it is nice to make sure it can be enforced in some way... but yeah, it could be made a MUST, or the language could be tweaked.

> can evaluate the list of rules from the first in the array to the last in the array, and attempt
> using the matching proxy or proxies from the earlies matching rule first. Multiple rules can

Reviewer 2: ```suggestion
can evaluate the list of rules from the first in the array to the last in the array, and attempt
using the matching proxy or proxies from the earliest matching rule first. If earliest matching
rule has empty list of `proxies` client SHOULD NOT send matching traffic to any proxy defined
in this PvD. Multiple rules can
```

> Rules in the `proxy-match` list SHOULD be provided in order of priority, such that a client
> can evaluate the list of rules from the first in the array to the last in the array, and attempt
> using the matching proxy or proxies from the earlies matching rule first. Multiple rules can
> match for the same destination, in which case all are considered to be accessible through the
> matching proxies in case the sets of proxies are different.

Reviewer 2: ```suggestion
Rules in the `proxy-match` array define destination-specific proxy selection logic. When determining which proxy or proxies to use for a given destination, the client MUST select the most specific applicable rule. Specificity is determined based on the type and granularity of the match criteria, according to the following precedence rules:

- Rules containing a `domains` key take precedence over rules containing only `subnets`.
- Within `domains`-based rules, more specific domain suffixes (i.e., longer matching suffixes) take precedence.
- Within `subnets`-based rules, more specific prefixes (i.e., longer prefix lengths) take precedence.
- Among rules with equal domain or subnet specificity, rules that include matching `ports` key take precedence over those that do not.
- Among rules that include a matching `ports` key, a rule matching an exact port number is preferred over a rule matching a broader port range.

If no rule with a matching `domains`, `subnets`, or `ports` key applies, the client SHOULD fall back to a rule with no such keys (i.e., an empty match criteria object), which acts as a default rule.
```

In order to match a destination rule in the `proxy-match` list, all properties MUST apply.

> A matched rule will then point to one or more proxy `identifier` values, which correspond
> to proxies defined in the list from {{proxy-enumeration}}.

Reviewer 2: ```suggestion
A matched rule will then either point to one or more proxy `identifier` values, which correspond
to proxies defined in the list from {{proxy-enumeration}}, or instructs the client to not send the
matching traffic to any proxy.
```

> Rules in the `proxy-exclude` list define properties of destinations that cannot be proxied
> either by a specific set of proxies, or all proxies. If the `proxies` key of the destination
> rule is present, then only the proxies listed cannot support the destination traffic. If the
> `proxies` key is missing, then none of the proxies associated with this PvD can support
> the destination traffic. As with match rules, in order to match an exclude rule, all properties
> MUST apply.
> 
> Exclude rules always take priority over rules in the `proxy-match` list. Clients SHOULD NOT

Reviewer 2: ```suggestion
Destination rules that do not contain `proxies` key define properties of destinations that
cannot be proxied by any proxy associated with this PvD. Clients SHOULD NOT send matching
traffic to any proxy associated with this PvD unless more specific matching destination rule with
`proxies` key exists.
```

> Rules in the `proxy-exclude` list define properties of destinations that cannot be proxied
> either by a specific set of proxies, or all proxies. If the `proxies` key of the destination
> rule is present, then only the proxies listed cannot support the destination traffic. If the
> `proxies` key is missing, then none of the proxies associated with this PvD can support
> the destination traffic. As with match rules, in order to match an exclude rule, all properties
> MUST apply.
> 
> Exclude rules always take priority over rules in the `proxy-match` list. Clients SHOULD NOT

Reviewer 2: ```suggestion
```

> Exclude rules always take priority over rules in the `proxy-match` list. Clients SHOULD NOT

Reviewer 3: This means that certain allows/excludes will be impossible to express. For example, if you want to allow 10.0.0.0/8, exclude 10.0.0.0/16, but allow 10.0.0.0/24, that's impossible.

You could say that more specific rules take priority (e.g., more specific subdomain or more specific subnet), and a tie breaks in favor of exclude.

Author: Options:
- Make more specific rules take priority to break ties
- Have a unified list where some rules say "no proxy", and the entire list is in priority order

> send traffic to a proxy if the destination matches a rule in the `proxy-exclude` list.

Reviewer 2: ```suggestion
```

> send traffic to a proxy if the destination matches a rule in the `proxy-exclude` list.

Reviewer 2: ```suggestion
```

has a local policy to only send requests for "*.example.com" to a proxy
"proxy.example.com", and `domains` array of a `match` object contains "internal.example.com" and
"other.company.com", the client would end up only proxying "internal.example.com"
through the proxy.

## Examples

In the following example, two proxies are defined with a common identifier, and
there is a single match rule for "*.internal.example.org".

~~~
{
  "identifier": "proxy.example.org.",
  "expires": "2023-06-23T06:00:00Z",
  "prefixes": [],
  "proxies": [
    {
      "protocol": "http-connect",
      "proxy": "proxy.example.org:80",
      "identifier": "default_proxy"
    },
    {
      "protocol": "connect-udp",
      "proxy": "https://proxy.example.org/masque{?target_host,target_port}",
      "identifier": "default_proxy"
    }
  ],
  "proxy-match": [
    {
      "domains": [ "*.internal.example.org" ],
      "proxies": [ "default_proxy" ]
    }
  ]
}
~~~

The client could then choose to use either proxy associated with the "default_proxy" identifier
for accessing names that fall within the "*.internal.example.org" zone. This would include the

> hostnames "internal.example.com", "foo.internal.example.org", etc.

Reviewer 2: ```suggestion
hostnames "internal.example.org", "foo.internal.example.org", "www.bar.internal.example.org" and all other hosts within "internal.example.org".
```

Corrected zone of the first hostname and added "deeper" FQDN to confirm that matching is not limited in depth.

  "proxy": "https://proxy.example.org/masque{?target_host,target_port}"
} ]

### proxy-match Key

JSON Key: proxy-match

Description: Array of proxy match rules, as dictionaries, associated with
entries in the `proxies` list.

Type: Array of dictionaries

Example: [ {

} ]

> ### proxy-exclude Key
> 
> JSON Key: proxy-exclude
> 
> Description: Array of proxy exclude rules, as dictionaries, associated with
> entries in the `proxies` list.
> 
> Type: Array of dictionaries
> 
> Example: [ {
> 
> } ]
> 

Reviewer 2: ```suggestion
```

New assignments in the "Proxy Protocol PvD Values" registry will be administered by IANA through Expert Review {{!RFC8126}}.
Experts are requested to ensure that defined keys do not overlap in names or semantics, and have clear format definitions.
The reference and notes fields MAY be empty.

## New PvD Proxy Destination Rule Registry {#proxy-info-iana}

IANA is requested to create a new registry "Proxy Destination Rule PvD Keys", within the "Provisioning Domains (PvDs)" registry page.

> This new registry reserves JSON keys for use in sub-dictionaries under the `proxy-match` and `proxy-exclude` key.

Reviewer 2: ```suggestion
This new registry reserves JSON keys for use in sub-dictionaries under the `proxy-match` key.
```
